### PR TITLE
Device enhance

### PR DESCRIPTION
**Pull request by uniartisan**
## What does this PR do?

Thank you to the lightning team for providing such an easy-to-use, clearly designed library.

The pr draft hopes to provide better registration support for a wide range of third-party hardware, and the pr is designed to integrate third-party hardware with minimal intrusive changes, including intel XPU, Huawei Ascend NPU, Cambrian, Moorethreads, and more.

Fixes #\<issue_number>

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

**Comment by uniartisan** [comment]
Examples here: https://github.com/uniartisan/RWKV-PEFT/blob/device-enhance/train.py#L499

There are a lot of things to be checked, I will try to do it later and make it more clear in documentation

File: src/lightning/fabric/strategies/ddp.py [hunk 1]
```diff
@@ -124,7 +124,13 @@ def setup_module(self, module: Module) -> DistributedDataParallel:
         """Wraps the model into a :class:`~torch.nn.parallel.distributed.DistributedDataParallel` module."""
         device_ids = self._determine_ddp_device_ids()
         # https://pytorch.org/docs/stable/notes/cuda.html#id5
-        ctx = torch.cuda.stream(torch.cuda.Stream()) if device_ids is not None else nullcontext()
+        ctx = (
+            getattr(torch, f"{self.root_device.type.split(':')[0]}").stream(
```
Comment (lantiga) [inline src/lightning/fabric/strategies/ddp.py:128]: I'd avoid doing this inline. Better to `getattr` and assign to a variable, and then use the variable.

My main question here is what is the contract that ensures that an accelerator has a concept of streams. Unless I read through the code, as a developer I wouldn't know that I need to register streams as `torch.mygpu.stream` and `torch.mygpu.Stream()`.

So we should either guard the strategy to only apply to `"cuda"`, or introduce a stream contract to the accelerator. I'd much rather do the former.

File: src/lightning/fabric/strategies/deepspeed.py [hunk 2]
```diff
@@ -507,7 +507,11 @@ def load_checkpoint(
 
         optimzer_state_requested = any(isinstance(item, (Optimizer, DeepSpeedOptimizer)) for item in state.values())
 
-        torch.cuda.empty_cache()
+        if isinstance(self.accelerator, Accelerator) and self.accelerator.get_device() != "cpu":
+            getattr(torch, self.root_device.type.split(":")[0]).empty_cache()
```
Comment (lantiga) [inline src/lightning/fabric/strategies/deepspeed.py:511]: Same goes for the comment above. `empty_cache` is not part of a contract (nor is the fact that a device is registered as a submodule of the `torch` module). It needs to be if we want to rely on calling `emtpy_cache` on whatever device we pass.

BTW, being a `torch` submodule is too strong of a requirement in my opinion.

In this case we should probably guard the strategy to be GPU-only.

**Review by lantiga — COMMENTED** [review]
Thank you for the interesting PR! I added a few comments.

**Comment by lantiga** [comment]
hey @uniartisan are you willing to finish this one up? It would be a welcome contribution

File: src/lightning/fabric/plugins/precision/amp.py [hunk 3]
```diff
@@ -51,7 +51,14 @@ def __init__(
 
         self.precision = precision
         if scaler is None and self.precision == "16-mixed":
-            scaler = torch.amp.GradScaler(device=device) if _TORCH_GREATER_EQUAL_2_4 else torch.cuda.amp.GradScaler()
+            scaler = (
+                torch.amp.GradScaler(device=device)
+                if _TORCH_GREATER_EQUAL_2_4
+                else getattr(
+                    torch,
+                    "cuda" if not isinstance(device, str) or device.split(":")[0] == "cpu" else device.split(":")[0],
```
[the remark this answers is not in the document]
Comment (uniartisan) [reply src/lightning/fabric/plugins/precision/amp.py:59]: Here I think I can improve it and use torch.device.type. The problem here is that if it is a cpu backend it will lack this function, so I use cuda instead

File: src/lightning/fabric/strategies/ddp.py (hunk 1 above)
Comment (uniartisan) [reply src/lightning/fabric/strategies/ddp.py:128]: Since I see that accelerators such as Intel and MUSA (some GPGPU hardware) support streams, I think it is possible to provide such support for multiple backends. If torch.mygpu.Stream() is not supported, it can fall back to a non-stream implementation. I will improve this in my next commit.

File: src/lightning/fabric/strategies/deepspeed.py (hunk 2 above)
Comment (uniartisan) [reply src/lightning/fabric/strategies/deepspeed.py:511]: Sorry, I may not understand what you mean. Could you tell me more about your idea?

**Comment by fritol** [comment]
will you add DirectML?

**Comment by uniartisan** [comment]
> will you add DirectML?

What I've added is the code for plug-in registration, which means you can register Direct ML by yourself. Just write a few simple function 🤓

**Comment by lantiga** [comment]
Thank you @uniartisan!

**Comment by zhiyuan1i** [comment]
@lantiga Hello, since pytorch 2.3, it has started to support Intel's SYCL post-computing device, alias torch.xpu, so I wonder if we can promote the landing of this PR. Is there anything I can do?
